WebExtensionContext::tabsUpdate and windowsUpdate need to properly capture lambdas.

xeenon opened the pull request:
#### 8db235e14a7fcce78fc3896d18b2259f7cf75406
<pre>
WebExtensionContext::tabsUpdate and windowsUpdate need to properly capture lambdas.
<a href="https://webkit.org/b/270804">https://webkit.org/b/270804</a>
<a href="https://rdar.apple.com/problem/124393562">rdar://problem/124393562</a>

Reviewed by Jeff Miller.

Fix tabsUpdate, windowsUpdate, and tabsRemove to properly capture the lambdas’
variables so the completion handler blocks can be called async. Also fixed
alarmsCreate to protect the WebExtensionContext lambda capture with a Ref.

Updated the TestWebExtensionTab and TestWebExtensionWindow to call all the completion
handlers async so this is tested.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
(WebKit::WebExtensionContext::alarmsCreate):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsUpdate):
(WebKit::WebExtensionContext::tabsRemove):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm:
(WebKit::WebExtensionContext::windowsUpdate):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab toggleReaderModeForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab detectWebpageLocaleForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab reloadForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab reloadFromOriginForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab goBackForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab goForwardForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab setParentTab:forWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab pinForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab unpinForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab muteForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab unmuteForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab duplicateForWebExtensionContext:withOptions:completionHandler:]):
(-[TestWebExtensionTab activateForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab selectForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab deselectForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab closeForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow setWindowState:forWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow setFrame:forWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow closeForWebExtensionContext:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/280233@main">https://commits.webkit.org/280233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34a9453ed05939a2fd9d151fc762b8f005d56ebd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56011 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35335 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59010 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6452 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58137 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42956 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6639 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45117 "Found 1 new test failure: http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4465 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33233 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48298 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26249 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30014 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5627 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4586 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51986 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60598 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6028 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52543 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48367 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52073 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8306 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31164 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32248 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33334 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31997 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jeffmapple reviewed (approved):
I assume the places you changed to dispatch back to the main queue are because the completion handlers can be called asynchronously? You might want to call this out in the commit message.

elijahsawyers commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
```diff
@@ -44,7 +44,7 @@
 void WebExtensionContext::alarmsCreate(const String& name, Seconds initialInterval, Seconds repeatInterval)
 {
-    m_alarmMap.set(name, WebExtensionAlarm::create(name, initialInterval, repeatInterval, [&](const WebExtensionAlarm& alarm) {
+    m_alarmMap.set(name, WebExtensionAlarm::create(name, initialInterval, repeatInterval, [this, protectedThis = Ref { *this }](const WebExtensionAlarm& alarm) {
```
What's the difference between `this` and `protectedThis`?

xeenon replied:
`this` is implicitly used by the member function call in the block. The `protectedThis` holds a reference to keep `this` alive, since it won't be retained otherwise. Declaring both in the capture lets the implicit use still work with out needing to do `protectedThis->foo()`.

xeenon committed:
WebExtensionContext::tabsUpdate and windowsUpdate need to properly capture lambdas.
https://webkit.org/b/270804
rdar://problem/124393562

Reviewed by Jeff Miller.

Fix tabsUpdate, windowsUpdate, and tabsRemove to properly capture the lambdas’
variables so the completion handler blocks can be called async. Also fixed
alarmsCreate to protect the WebExtensionContext lambda capture with a Ref.

Updated the TestWebExtensionTab and TestWebExtensionWindow to call all the completion
handlers async so this is tested.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
(WebKit::WebExtensionContext::alarmsCreate):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsUpdate):
(WebKit::WebExtensionContext::tabsRemove):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm:
(WebKit::WebExtensionContext::windowsUpdate):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab toggleReaderModeForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab detectWebpageLocaleForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab reloadForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab reloadFromOriginForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab goBackForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab goForwardForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab setParentTab:forWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab pinForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab unpinForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab muteForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab unmuteForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab duplicateForWebExtensionContext:withOptions:completionHandler:]):
(-[TestWebExtensionTab activateForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab selectForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab deselectForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionTab closeForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow setWindowState:forWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow setFrame:forWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow closeForWebExtensionContext:completionHandler:]):

Canonical link: https://commits.webkit.org/280233@main
